### PR TITLE
qemu: Update to QEMU 6.2 release

### LINKED ISSUE #420: Update Qemu to get MVE support for the MPS3 AN547
The mps3 an547, a cortex-m55 cpu with MVE support, fails in the `tests/lib/sprintf/libraries.libc.sprintf` test with incorrect printing. This is caused by incorrectly decoding added-by-MVE shifts as or with PC as a source register. With these shifts decoding as or, the divide by 5 and divide by 10 routines don't work. This has been fixed in upstream qemu, probably related to https://www.mail-archive.com/qemu-devel@nongnu.org/msg827164.html

### LINKED ISSUE #423: Update QEMU to support GICv3 ITS for AARCH64 targets
Current version of toolchain 0.13 doesn't support GICv3 ITS emulation because the support was in review when the corresponding Qemu relase was used.

Since Qemu 6.2, GICv3 ITS emulation is supported for AARCH64 targets, and will permit enabling automatic CI testing for https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/arch/arm64/arm64_gicv3_its introduced in https://github.com/zephyrproject-rtos/zephyr/pull/37506.

And it will also permit adding test for pending https://github.com/zephyrproject-rtos/zephyr/pull/40566 and https://github.com/zephyrproject-rtos/zephyr/pull/40720 regarding MSI/MSI-X & PCIe bridges support

### PR DESCRIPTION
**Pull request by stephanosio**
This commit updates the QEMU version to 6.2.

QEMU 6.2 is required to support ARM M-Profile Vector Extension (MVE)
emulation.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

Closes #420.
Closes #423.

### CODE REVIEW

File: meta-zephyr-sdk/recipes-devtools/qemu/zephyr-qemu_git.bb
Comment (galak) [inline meta-zephyr-sdk/recipes-devtools/qemu/zephyr-qemu_git.bb:10]: Seems like some of these patches should drop as I think they got back ported from qemu-6.2.

Should we move go zephyr git repo for qemu?

**Review by galak — DISMISSED** [review]
feels like the patch list needs modification.

File: meta-zephyr-sdk/recipes-devtools/qemu/zephyr-qemu_git.bb
Comment (stephanosio) [reply meta-zephyr-sdk/recipes-devtools/qemu/zephyr-qemu_git.bb:10]: Updated to use the [`zephyr-qemu-v6.2.0` branch](https://github.com/zephyrproject-rtos/qemu/tree/zephyr-qemu-v6.2.0) in the Zephyr QEMU fork repository.

**Comment by stephanosio** [comment]
DO NOT MERGE: Validating the commit that switches to using the Zephyr fork first. The commit that actually updates QEMU to v6.2.0 will be added here soon.

**Comment by stephanosio** [comment]
**Patch Status**
```
0001-hw-misc-mps2-scc-Add-QEMU-interface-comment.patch
  Merged upstream (c52c266d24b10f1482602e6d22938d9e21f874f5)
0002-hw-misc-mps2-scc-Support-using-CFG0-bit-0-for-remapp.patch
  Merged upstream (5bddf92e689c0a3da57f4fd17b83d4eb1e436b80)
0003-hw-arm-mps2-tz-Implement-AN524-memory-remapping-via-.patch
  Merged upstream (f1dfab0d9b77b8649d60ded3d96870f88d0baccd)
0004-qemu-nios2-Add-Altera-MAX-10-board-support-for-Zephy.patch
  Applied as-is (Zephyr-specific)
0005-hw-sparc-Add-leon-at697-machine.patch
  Applied as-is
0006-hw-sparc-leon-Fix-compilation-errors.patch
  Applied as-is
0007-hw-sparc-leon-timer-Call-leon_timer_io_read-for-TIME.patch
  Applied as-is
0008-hw-sparc-leon-Switch-to-transaction-based-ptimer-API.patch
  Applied as-is
0009-hw-sparc-leon-Fix-compilation-errors-with-qemu-6-cha.patch
  Applied as-is
0010-Add-ARC-support.patch
  Not applied (Multiple conflicts, patch too big to be reworked)
0011-hw-arm-mps2-tz-Don-t-duplicate-modelling-of-SRAM-in-.patch
  Merged upstream (b6889c5ae3895cf5a4322adb32b2133e9b91d158)
0012-hw-arm-mps2-tz-Make-SRAM_ADDR_WIDTH-board-specific.patch
  Merged upstream (902b28ae4eba6df303cba57016945426865a6d59)
0013-hw-arm-armsse.c-Correct-modelling-of-SSE-300-interna.patch
  Merged upstream (4eb1770988397354ed95dec2bd63c09348ebf707)
0014-hw-arm-armsse-Convert-armsse_realize-to-use-ERRP_GUA.patch
  Merged upstream (32962103523838faf15c070bff0fc75dfd2f42c8)
0015-hw-arm-mps2-tz-Allow-board-to-specify-a-boot-RAM-siz.patch
  Merged upstream (2f12dca05928078ecc5c7d209a2bc5af61bff966)
0016-hw-arm-Model-TCMs-in-the-SSE-300-not-the-AN547.patch
  Merged upstream (cbb563887781d813e67c59b68dd76891cb78c3d4)
0017-target-arm-Use-correct-SP-in-M-profile-exception-ret.patch
  Merged upstream (659f042ba8ef80ca1053042a781a4488f9a587dc)
0018-arc-virt-Make-target-memory-size-configurable.patch
  Not applied (Depends on '0010-Add-ARC-support.patch')
```

**Comment by stephanosio** [comment]
@abrodkin It looks like the QEMU ARC patch (`0010-Add-ARC-support.patch`) has not been upstreamed yet. Are you able to provide an updated patch based on top of the QEMU 6.2 release?

**Comment by abrodkin** [comment]
@stephanosio well my patch could be easily applied on top of QEMU v6.2 in theory. But the reality is ARC port is still based on upstream v6.0.0 and my quick attempt to rebase on v6.2.0 was not successful as there're not only build-system changes but also some data structures etc. That said I'd like our QEMU guru @cupertinomiranda to conduct that re-base and since he's on Christmas holidays it might only happen sometime early next year.

And given ARC pieces were not yet accepted into upstream QEMU I'd like to ask to wait a bit for QEMU update in Zephyr SDK, as otherwise all ARC support will be effectively dropped from QEMU in the SDK/perf-PR CI etc. And I'd like to not add `qemu-arc` to the `meta-zephyr-sdk` by all means (at least now).

**Comment by stephanosio** [comment]
@abrodkin Thanks. I will put this PR on hold until the rebased ARC patch is provided.

**Comment by abrodkin** [comment]
@stephanosio just in case when do you think you'll really need to switch to v6.2? To make sure because of other things we don't slip way too much our rebase.

**Comment by stephanosio** [comment]
> just in case when do you think you'll really need to switch to v6.2?

@abrodkin Ideally, as soon as possible. The motivation for switching to v6.2 is to support the ARM MVE instruction emulation and testing in the CI using the QEMU (see #420). There are a few Zephyr PRs that are waiting for this as well.

**Comment by stephanosio** [comment]
TODO: Install/update meson version in the SDK CI image (`pip3 install meson`).

```
../git/meson.build:1:0: ERROR: Meson version is 0.55.1 but project requires >=0.58.2
```

**Comment by stephanosio** [comment]
@abrodkin @cupertinomiranda any updates?

**Comment by abrodkin** [comment]
@stephanosio thanks to the work done by @cupertinomiranda we've got our port re-based on top of v6.2.0.
Now I'm going to test it against Zephyr and once I see it works as good as before and will send a PR for the SDK-ng.

**Comment by stephanosio** [comment]
> @stephanosio thanks to the work done by @cupertinomiranda we've got our port re-based on top of v6.2.0. Now I'm going to test it against Zephyr and once I see it works as good as before and will send a PR for the SDK-ng.

@abrodkin Thanks! When you are done testing, you can open a pull request to https://github.com/zephyrproject-rtos/qemu/commits/zephyr-qemu-v6.2.0, or you can provide a patch file, whichever is more convenient for you.

**Comment by microbuilder** [comment]
Just a heads up that there is a bug with systick in QEMU 6.2.0 that breaks mps2_an521 and mps3_an547 support, perhaps others. I tested this fix this morning, but we will either need to update to the next release to fix MPSx support, or apply this patch: https://lists.gnu.org/archive/html/qemu-devel/2022-02/msg00333.html

**Comment by stephanosio** [comment]
@microbuilder Thanks for the heads up. I will make sure to apply that patch and test all supported ARM platforms before merging this.

**Comment by microbuilder** [comment]
> Thanks for the heads up. I will make sure to apply that patch and test all supported ARM platforms before merging this.

An LSM303DLHC magnetometer driver I wrote was merged in, as well as a change to make a subset of I2C drivers available to most ARM QEMU targets that enable I2C (mps2_an521, mps3_an547, etc.). This is after the 6.2.0 tag, but I think these changes are valuable to Zephyr to enable testing I2C in CI with QEMU.

- https://lists.gnu.org/archive/html/qemu-devel/2022-02/msg01482.html
- https://lists.gnu.org/archive/html/qemu-devel/2022-02/msg01818.html

You can enable this optional driver with something like:

```
if(BOARD MATCHES "mps2_an521")
  list(APPEND QEMU_EXTRA_FLAGS -device lsm303dlhc_mag,id=lsm303,address=0x1E)
endif()
```

**Comment by abrodkin** [comment]
@stephanosio oh, sorry for the delay. Indeed, we have now usable QEMU port for ARC based on 6.2.
The reason I was a bit hesitant on posting an update - there's one reliable regression now: https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/kernel/timer/timer_monotonic fails on all ARC QEMU targets.

And even though we did and still plan to solve it, to not delay SDK update with newer QEMU, I guess it should be fine to just ignore that test for ARC QEMU platforms for a time being.

**Comment by stephanosio** [comment]
@abrodkin Thanks for the update. Since it looks like it will take some time to upstream the ARC support for QEMU and it will be hard to keep the Zephyr SDK QEMU version and the ARC QEMU version in sync until that happens, I think it would be best to create a separate ARC QEMU component, which builds directly from https://github.com/foss-for-synopsys-dwc-arc-processors/qemu, as we do with the Xilinx QEMU.

What do you think?

**Comment by stephanosio** [comment]
Postponing this to 0.14.1, which will be a quick follow-up to 0.14.0 (multi-platform SDK) in a month or so.

**Comment by abrodkin** [comment]
@stephanosio if there're no generic Zephyr-specific patches on top of vanilla QEMU, it might indeed be a good idea to build right from our source tree. If will be much simpler in a sense of logistics as well as less cross-dependencies.
Let me submit a patch for the SDK in the same way as it's done for Xtensa.

**Comment by stephanosio** [comment]
> if there're no generic Zephyr-specific patches on top of vanilla QEMU

@abrodkin Most patches we have for QEMU have to do with ARM, so building ARC QEMU directly from the Synopsys repo should be ok.

**Comment by stephanosio** [comment]
The following ARC-related patches will not be integrated here since we will be providing a separate ARC QEMU build:
```
0010-Add-ARC-support.patch
0018-arc-virt-Make-target-memory-size-configurable.patch
```

**Comment by stephanosio** [comment]
> Just a heads up that there is a bug with systick in QEMU 6.2.0 that breaks mps2_an521 and mps3_an547 support, perhaps others. I tested this fix this morning, but we will either need to update to the next release to fix MPSx support, or apply this patch: https://lists.gnu.org/archive/html/qemu-devel/2022-02/msg00333.html

@microbuilder I have integrated this patch here and verified that the Zephyr kernel tests pass on the MPS boards.

**Comment by stephanosio** [comment]
> An LSM303DLHC magnetometer driver I wrote was merged in, as well as a change to make a subset of I2C drivers available to most ARM QEMU targets that enable I2C (mps2_an521, mps3_an547, etc.). This is after the 6.2.0 tag, but I think these changes are valuable to Zephyr to enable testing I2C in CI with QEMU.
> 
> * https://lists.gnu.org/archive/html/qemu-devel/2022-02/msg01482.html
> * https://lists.gnu.org/archive/html/qemu-devel/2022-02/msg01818.html

@microbuilder I will look into integrating these patches in a separate PR.

**Comment by stephanosio** [comment]
@abrodkin I have verified that all ARC tests pass using the separate ARC QEMU build here.

**Comment by stephanosio** [comment]
@microbuilder @abrodkin please review

File: meta-zephyr-sdk/recipes-devtools/arc_qemu/files/cross.patch
Comment (abrodkin) [inline meta-zephyr-sdk/recipes-devtools/arc_qemu/files/cross.patch:13]: Offsets differ from the same patch for generic QEMU, so maybe you copy over one version over another so that they match exactly? Otherwise it might look a bit suspicious.

**Review by abrodkin — APPROVED** [review]
Awesome! Thanks for all the work here. LGTM!
The only thing which is IMHO missing inability to download built artifacts, i.e. it's not possible to test-drive built entire SDK. Maybe it's due to this repo settings as back in the days of Buildkite I was able to do so (thanks to @galak who granted me extended permissions).

**Comment by stephanosio** [comment]
> The only thing which is IMHO missing inability to download built artifacts, i.e. it's not possible to test-drive built entire SDK. Maybe it's due to this repo settings as back in the days of Buildkite I was able to do so (thanks to @galak who granted me extended permissions).

@abrodkin You should find the artifacts under the checks "Summary" tab:
https://github.com/zephyrproject-rtos/sdk-ng/actions/runs/2000487547

**Comment by abrodkin** [comment]
@stephanosio ooops, I didn't notice that section as it was just below my screen edge, sorry for that!

**Comment by microbuilder** [comment]
@stephanosio It looks like we only generate x86 builds? I'm on an AArch64 machine, but I'll setup an x86-64 machine to test these and make sure QEMU behaves as expected with the MPS2 and MPS3 targets, etc.

**Comment by stephanosio** [comment]
> @stephanosio It looks like we only generate x86 builds? I'm on an AArch64 machine, but I'll setup an x86-64 machine to test these and make sure QEMU behaves as expected with the MPS2 and MPS3 targets, etc.

PR builds are, by default, x86-64 only (need to label the PR `ci-linux-aarch64` to build for AArch64).

**Comment by microbuilder** [comment]
@stephanosio We'll need this patch in the Zephyr SDK or the systick timer is broken on a number of platforms we use to test Arm in CI: https://patchew.org/QEMU/20220201192650.289584-1-git@rpls.de/

The issue was introduced between 6.1.0 and 6.2.0.

**Comment by stephanosio** [comment]
> @stephanosio We'll need this patch in the Zephyr SDK or the systick timer is broken on a number of platforms we use to test Arm in CI: https://patchew.org/QEMU/20220201192650.289584-1-git@rpls.de/
> 
> The issue was introduced between 6.1.0 and 6.2.0.

Not sure if you saw https://github.com/zephyrproject-rtos/sdk-ng/pull/421#issuecomment-1071883338. I have already applied that patch and tested that all tests for the emulated platforms (including MPS) pass.

**Review by microbuilder — APPROVED** [review]

**Comment by abrodkin** [comment]
@stephanosio for the record with the SDK build from this PR's CI I have very clean results while executing full `twister` on all 3 ARC QEMU platforms: `qemu_arc_em`, `qemu_arc_hs` & `qemu_arc_hs6x` with just 1 failure, already captured in https://github.com/foss-for-synopsys-dwc-arc-processors/qemu/issues/67:
```
$ ./scripts/twister --platform qemu_arc_em --platform qemu_arc_hs --platform qemu_arc_hs6x
INFO    - Zephyr version: zephyr-v3.0.0-1286-g38938b4e594a
INFO    - JOBS: 12
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testcase list...
INFO    - 1651 test scenarios (4686 configurations) selected, 3483 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:  553/1203  45%  skipped: 3601, failed:    0
ERROR   - qemu_arc_em               tests/kernel/mem_protect/syscalls/kernel.memory_protection.syscalls FAILED: Exited with 2
ERROR   - see: .../twister-out/qemu_arc_em/tests/kernel/mem_protect/syscalls/kernel.memory_protection.syscalls/handler.log
INFO    - Total complete: 1203/1203  100%  skipped: 3851, failed:    1
INFO    - 834 of 835 test configurations passed (99.88%), 1 failed, 3851 skipped with 0 warnings in 3971.16 seconds
INFO    - In total 5425 test cases were executed, 55082 skipped on 3 out of total 442 platforms (0.68%)
INFO    - 706 test configurations executed on platforms, 129 test configurations were only built.
```

Again, thanks for all the efforts!
And shouldn't we close https://github.com/zephyrproject-rtos/sdk-ng/issues/445 now?

**Pull request merged**
